runtime/src/main/scala/akka/grpc/internal/AkkaHttpClientUtils.scala

```diff
@@ -0,0 +1,226 @@
+          res.attribute(ResponsePromise.Key).get.promise.trySuccess(res)
+        })(Keep.both)
+        .run()
+
+    def singleRequest(request: HttpRequest): Future[HttpResponse] = {
+      val p = Promise[HttpResponse]()
+      queue
+        .offer(request.addAttribute(ResponsePromise.Key, ResponsePromise(p)))
+        .flatMap(o => {
+          log.info(s"XXX enqueue: $o")
+          p.future.onComplete(t => log.info(s"XXX dequeue: ${t.isSuccess}"))
+          p.future
+        })
+    }
+
+    implicit def serializerFromMethodDescriptor[I, O](descriptor: MethodDescriptor[I, O]): ProtobufSerializer[I] =
+      descriptor.getRequestMarshaller.asInstanceOf[WithProtobufSerializer[I]].protobufSerializer
+    implicit def deserializerFromMethodDescriptor[I, O](descriptor: MethodDescriptor[I, O]): ProtobufSerializer[O] =
+      descriptor.getResponseMarshaller.asInstanceOf[WithProtobufSerializer[O]].protobufSerializer
+
+    new InternalChannel() {
+      override def shutdown(): Unit = queue.complete()
+      override def done: Future[Done] = doneFuture
+
+      override def invoke[I, O](
+          request: I,
+          headers: MetadataImpl,
+          descriptor: MethodDescriptor[I, O],
+          options: CallOptions): Future[O] = {
+        val src =
+          invokeWithMetadata(Source.single(request), descriptor.getFullMethodName, headers, descriptor, false, options)
+        val result = src.toMat(Sink.headOption)(Keep.right).run()
+        result.flatMap {
+          case Some(r) => Future.successful(r)
+          case None    => Future.failed(throw new IllegalStateException("Successful status code but no data found"))
+        }
+      }
+
+      override def invokeWithMetadata[I, O](
+          request: I,
+          headers: MetadataImpl,
+          descriptor: MethodDescriptor[I, O],
+          options: CallOptions): Future[GrpcSingleResponse[O]] = ???
+
+      override def invokeWithMetadata[I, O](
+          source: Source[I, NotUsed],
+          fqMethodName: String,
+          headers: MetadataImpl,
+          descriptor: MethodDescriptor[I, O],
+          streamingResponse: Boolean,
+          options: CallOptions): Source[O, Future[GrpcResponseMetadata]] = {
+        implicit val serializer: ProtobufSerializer[I] = descriptor
+        val deserializer: ProtobufSerializer[O] = descriptor
+        val httpRequest = GrpcRequestHelpers(
+          Uri(s"https://${settings.overrideAuthority.getOrElse(host)}/" + descriptor.getFullMethodName),
+          source)
+        Source.lazyFutureSource[O, Future[GrpcResponseMetadata]](() => {
+          singleRequest(httpRequest).map { response =>
+            {
+              Codecs.detect(response) match {
+                case Success(codec) =>
+                  log.info(s"XXX response $response started")
+                  implicit val reader: GrpcProtocolReader = GrpcProtocolNative.newReader(codec)
+                  val trailerPromise = Promise[immutable.Seq[HttpHeader]]()
```

Comment:
Since the `trailerPromise` is not always related to the trailers (it may be redeemed if the upstream is completed) maybe we should have a comment or rename it?

(this is more relevant on #1179)